sandbox/cgroup/monitor.go

```diff
@@ -33,122 +33,190 @@ import (
-	// InDelete one, so we must monitor the parent folder and detect when a
-	// specific folder, corresponding to a CGroup, has been deleted.
-	// Each path has associated an integer, which is an usage counter. Every
-	// time a new CGroup paths have to be monitored, the parent paths are
-	// checked in the map, and if they already exist, the usage counter is
-	// incremented by one (of course, if a parent path isn't in the map,
-	// it is initialized to one). Every time the path of a CGroup is deleted,
-	// the corresponding usage counter here corresponding to the parent path
-	// is decremented, and when it reaches zero, it is removed and inotify is
-	// notified to stop monitoring it.
-	pathList map[string]int32
+	// The paths monitored are both the actual leaf path and the parent
+	// directory. See discussion in addWatch for details on why the parent
+	// needs to be tracked as well.
+	pathList map[string]uint
+
+	done     func()
+	doneChan chan struct{}
+
+	// observer callback to facilitate testing, called when a watch is
+	// added, or a folder from a watched group is removed, or the whole
+	// group is empty and thus notified about
+	observeMonitorCb func(w *inotifyWatcher, name string)
 }
 
 // Contains the data corresponding to a CGroup that must be watched to detect
 // when it is destroyed
 type groupToWatch struct {
 	// This is the CGroup identifier
 	name string
-	// This contains a list of folders to monitor. When all of them have been
+	// This contains a hash map of folders to monitor. When all of them have been
 	// deleted, the CGroup has been destroyed and there are no processes running
-	folders []string
+	folders map[string]struct{}
 	// This channel is used to notify to the requester that this CGroup has been
 	// destroyed. The watcher writes the CGroup identifier on it; this way, the
 	// same channel can be shared to monitor several CGroups.
 	channel chan<- string
 }
 
-var currentWatcher *inotifyWatcher = &inotifyWatcher{
-	wd:           nil,
-	pathList:     make(map[string]int32),
-	addWatchChan: make(chan *groupToWatch),
+var currentWatcher *inotifyWatcher = newInotifyWatcher(context.Background())
+
+func newInotifyWatcher(ctx context.Context) *inotifyWatcher {
+	ctx, cancel := context.WithCancel(ctx)
+	return &inotifyWatcher{
+		ctx:          ctx,
+		pathList:     make(map[string]uint),
+		addWatchChan: make(chan *groupToWatch),
+		doneChan:     make(chan struct{}),
+		done:         cancel,
+	}
 }
 
 func (gtw *groupToWatch) sendClosedNotification() {
-	defer func() {
-		if err := recover(); err != nil {
-			logger.Noticef("Failed to send Closed notification for %s", gtw.name)
-		}
+	go func() {
+		defer func() {
+			if err := recover(); err != nil {
+				logger.Noticef("cannot send closed notification for %s", gtw.name)
+			}
+		}()
+		gtw.channel <- gtw.name
 	}()
-	gtw.channel <- gtw.name
+}
+
+// Close stops the watcher and waits for it to finish.
+func (iw *inotifyWatcher) Close() {
+	iw.done()
+	<-iw.doneChan
```

Comment:
I presume this waits for the watcher loop to terminate, right?

Reply:
yes